adr/2024-09-fork-safety.md

```diff
@@ -0,0 +1,57 @@
+
+# 2024-09 Implement `Database#discard` to mitigate sqlite's lack of fork safety
+
+## Status
+
+Proposed, seeking feedback and suggestions for other alternative solutions.
+
+
+## Context
+
+In August 2024, Andy Croll opened an issue[^issue] describing sqlite file corruption related to solid queue. After investigation, we were able to reproduce corruption under certain circumstances when forking a process with open sqlite databases.[^repro]
+
+SQLite is known to not be fork-safe[^howto], so this was not entirely surprising though it was the first time your author had personally seen corruption in the wild. The corruption became much more likely after the sqlite3-ruby gem improved its memory management with respect to open statements[^gemleak] in v2.0.0.
+
+Advice from upstream contributors[^advice] is, essentially: don't fork if you have open database connections. Or, if you have forked, don't call `sqlite3_close` on those connections (which will result in a memory leak in the child process). Neither of these options are great, see below.
+
+
+## Decision
+
+The sqlite3-ruby gem will introduce support for `Database#discard` in v2.1.0, which can be called by frameworks like Rails that have made the decision to allow forking while connections are open. This method will close the underlying file descriptors, but will not free any of the memory or call `sqlite3_close` to clean up internal data structures.
+
+
+## Consequences
+
+The positive consequence is that frameworks like Rails can prevent database file corruption as a result of allowing forks by simply `discard`ing the inherited connections in the child process.
+
+The negative consequence is that, when choosing this option, some memory will be permanently lost (leaked) in the child process.
+
+
+## Alternatives considered.
+
+### 1. Require applications to close database connections before forking.
+
+This is the advice[^advice] given by the upstream maintainers of sqlite, and so was the first thing we tried to implement in Rails in [rails/rails#52931](https://github.com/rails/rails/pull/52931)[^before_fork]. That first simple implementation was not thread safe, however, and in order to make it thread-safe it would be necessary to pause all sqlite database activity, close the open connections, and then fork. At least one Rails core team member was not happy that this would interfere with database connections in the parent, and the complexity of a thread-safe solution seemed high, so this work was paused.
```

Comment:
(sorry, just noticed now I queued the review comment instead of publishing it hours ago)

Reply:
> have all the connection share a read-write lock

In what way would the connections share this? What operations would be behind the read lock?

I started implementing something like this -- a global shared/exclusive lock owned by the SQLite3Adapter class -- and then had `ConnectionPool#checkout` grab the shared lock. But there were still race conditions and closed connections being passed around somewhere. It didn't seem straightforward to me, but I'm not that familiar with connection management code or I may have misunderstood your suggestion.

Reply:
If you want to take a look at what I started, it's here: https://github.com/flavorjones/rails/tree/flavorjones-sqlite3-fork-safety

Reply:
> In what way would the connections share this? 

Would basically be a global/constant, e.g.:

```ruby
module Sqlite3
  FORK_LOCK = Concurrent::ReadWriteLock .new
end
```

> What operations would be behind the read lock?

Anything requiring the connection not to be closed. So `query`, `transaction do`, etc.

> a global shared/exclusive lock owned by the SQLite3Adapter class

I was thinking of doing this in Sqlite3-ruby, not active record. Because the idea isn't to lock when a connection is checked out, but when it's in active use. This way if a thread is still holding a connection but not actiely using it, we can close + `sqlite3_free` under its nose. And then later when the fork lock is released and a thread try to use such connection again, we recreate it on the fly.

Of course there are probably complications I'm not thinking of, like prepared statements would be invalidated? etc.

Reply:
> prepared statements would be invalidated

Yes, this is what I found to be the problem on the branch I linked above.

Reply:
> I would love to chat with you and other folks (maybe at Rails World?)

Yeah, I'm sure we can find time with Matthew to chat about that. Dunno if there's really a good solution to this problem, but...